src/crimson/os/seastore/seastore_types.cc

```diff
@@ -239,20 +242,160 @@ std::string device_type_to_string(device_type_t dtype) {
 paddr_t convert_blk_paddr_to_paddr(blk_paddr_t addr, size_t block_size,
     uint32_t blocks_per_segment, device_id_t d_id)
 {
-  paddr_t paddr;
-  paddr.segment = segment_id_t {
+  segment_id_t id = segment_id_t {
     d_id,
-    (uint32_t)(addr / (block_size * blocks_per_segment))
+    (device_segment_id_t)(addr / (block_size * blocks_per_segment))
   };
-  paddr.offset = addr % (block_size * blocks_per_segment);
-  return paddr;
+  segment_off_t off = addr % (block_size * blocks_per_segment);
+  return paddr_t{id, off};
 }
 
 blk_paddr_t convert_paddr_to_blk_paddr(paddr_t addr, size_t block_size,
     uint32_t blocks_per_segment)
 {
-  return (blk_paddr_t)(addr.segment.device_segment_id() *
-	  (block_size * blocks_per_segment) + addr.offset);
+  seg_paddr_t* s = addr.as_seg_paddr();
+  return (blk_paddr_t)(s->get_segment_id().device_segment_id() *
+	  (block_size * blocks_per_segment) + s->get_segment_off());
+}
+
+paddr_t::paddr_t(device_id_t id, device_segment_id_t sgt, segment_off_t offset) {
+  static_cast<seg_paddr_t*>(this)->set_segment_id(segment_id_t{id, sgt});
+  static_cast<seg_paddr_t*>(this)->set_segment_off(offset);
+}
+
+const seg_paddr_t* paddr_t::as_seg_paddr() const {
+  assert(get_addr_type() == addr_types_t::SEGMENT);
+  return static_cast<const seg_paddr_t*>(this);
+}
+
+seg_paddr_t* paddr_t::as_seg_paddr() {
+  assert(get_addr_type() == addr_types_t::SEGMENT);
+  return static_cast<seg_paddr_t*>(this);
+}
+
+void paddr_t::set_device_id(device_id_t id, addr_types_t type) {
+  dev_addr &= static_cast<common_addr_t>(
+    std::numeric_limits<device_segment_id_t>::max());
+  dev_addr |= static_cast<common_addr_t>(id & 0x8) << DEV_ADDR_LEN_BITS;
+  dev_addr |= static_cast<common_addr_t>(type)
+    << (std::numeric_limits<common_addr_t>::digits - 1);
+}
+
+#define PADDR_OPERATION(a_type, base, func)        \
+  if (get_addr_type() == a_type) {                 \
+    return static_cast<const base*>(this)->func;   \
+  }
+
+paddr_t paddr_t::operator-(paddr_t rhs) const {
+  if (get_addr_type() == addr_types_t::SEGMENT) {
+    return ((*static_cast<const seg_paddr_t*>(this)) - rhs);
+  }
+  ceph_assert(0 == "not supported type");
+  return paddr_t{};
+}
+
```

Comment:
There are at the moment 7 special paddr types: max, null, fake, zero, delayed, record_relative, and block_relative

It seems to me that the first 4 (max, null, fake, zero) are singleton special values independent of addr type.  Let's give each of those specific paddr_t::dev_addr values.

record_relative, and block_relative are used for physical addresses within a journal record pointing back into that record.  They are useful for the current scheme not due to zns devices, but rather to avoid needing the final journal offset until write time --  this may also be useful for an RBM journal.  Let's give each of those 3 (record_relative, block_relative, and also delayed) a device_id_t instead of a segment_id_t.